scipy/spatial/transform/tests/test_rotation.py

```diff
@@ -658,39 +652,114 @@ def test_as_euler_degenerate_symmetric_axes():
+            mat_estimated = Rotation.from_euler(
+                seq, angle_estimates, degrees=True
+                ).as_matrix()
 
-                assert_array_almost_equal(mat_expected, mat_estimated)
+            assert_array_almost_equal(mat_expected, mat_estimated)
+
+
+def test_as_euler_compare_algorithms():
+    rnd = np.random.RandomState(0)
+    n = 10
+    angles = np.empty((n, 3))
+    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
+    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
 
-                # Intrinsic rotations
-                seq = seq.upper()
-                rotation = Rotation.from_euler(seq, angles, degrees=True)
-                mat_expected = rotation.as_matrix()
+    # asymmetric axes
+    angles[:, 1] = rnd.uniform(low=-np.pi / 2, high=np.pi / 2, size=(n,))
+    for seq_tuple in permutations('xyz'):
+        # Extrinsic rotations
+        seq = ''.join(seq_tuple)
+        rot = Rotation.from_euler(seq, angles)
+        assert_allclose(rot._as_euler_from_matrix(seq), rot.as_euler(seq))
+        # Intrinsic rotations
+        seq = seq.upper()
+        rot = Rotation.from_euler(seq, angles)
+        assert_allclose(rot._as_euler_from_matrix(seq), rot.as_euler(seq))
+        
+    # symmetric axes
+    angles[:, 1] = rnd.uniform(low=0, high=np.pi, size=(n,))
+    for seq_tuple in permutations('xyz'):
+        # Extrinsic rotations
+        seq = ''.join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
+        rot = Rotation.from_euler(seq, angles)
+        assert_allclose(rot._as_euler_from_matrix(seq), rot.as_euler(seq))
+        # Intrinsic rotations
+        seq = seq.upper()
+        rot = Rotation.from_euler(seq, angles)
+        assert_allclose(rot._as_euler_from_matrix(seq), rot.as_euler(seq))
 
-                angle_estimates = rotation.as_euler(seq, degrees=True)
-                mat_estimated = Rotation.from_euler(
-                    seq, angle_estimates, degrees=True
-                    ).as_matrix()
 
-                assert_array_almost_equal(mat_expected, mat_estimated)
+def test_as_euler_degenerate_compare_algorithms():
+    # asymmetric axes
+    angles = np.array([
+        [45, 90, 35],
+        [35, -90, 20],
+        [35, 90, 25],
+        [25, -90, 15]])
 
+    with pytest.warns(UserWarning, match="Gimbal lock"):
+        for seq_tuple in permutations('xyz'):
+            # Extrinsic rotations
+            seq = ''.join(seq_tuple)
+            rot = Rotation.from_euler(seq, angles, degrees=True)
+            assert_allclose(
+                rot._as_euler_from_matrix(seq, degrees=True)[:,[0,2]],
+                rot.as_euler(seq, degrees=True)[:,[0,2]])
+            
+            # Intrinsic rotations
+            seq = seq.upper()
+            rot = Rotation.from_euler(seq, angles, degrees=True)
+            assert_allclose(
+                rot._as_euler_from_matrix(seq, degrees=True)[:,[0,2]], 
```

Comment:
Here a comment is useful to explain why angle 1 is not compared. Can we compare it with looser tolerance?

There is a tolerance parameters in `assert_allclose`. Conceptually we wan't to make it as small as possible, to show that the equivalence is very tight. The same can be said for `test_as_euler_compare_algorithms`.

Reply:
The idea in general is to use as small tolerances as possible everywhere. At least for tests which are considered important. 

It means that angles[0] and angles[2] are tested with one tolerance parameters (and likely lower than the default) and angles[1] with another parameter.

The same logic for `test_as_euler_compare_algorithms`.

Reply:
`test_as_euler_compare_algorithms` does not really have any problems, since the numerical differences are bigger for only for `angles[1] = 0` or `180` for symmetric axes and `angles[1] = -90` or `90` for assymmetric axes.

I realized that I can make everything pass the test in `test_as_euler_compare_algorithms_degenerate` by changing the `atol` from `atol = 1e-07` to `atol = 1e-06`. Since it is such a small change (well, a whole order of magnitude but still pretty small), I decided to just apply it to everything so the test becomes a bit less awkward.

If you prefer it though, I can test `angles[0]` and `angles[2]` with the default value of `atol = 1e-07` and `angles[1]` with the default value of `atol = 1e-06`.

Reply:
>test_as_euler_compare_algorithms does not really have any problems, since the numerical differences are bigger for only for angles[1] = 0 or 180 for symmetric axes and angles[1] = -90 or 90 for assymmetric axes.

But ideally the results should be exactly the same. Can you check much lower tolerances? Also I think `rtol` should be tweaked here (i. e. tolerance relative to the value), because angles are typically are not close to zero.

That is start with `atol = 0` and `rtol = 1e-7` and then decrease `rtol` down while tests pass.

> If you prefer it though, I can test angles[0] and angles[2] with the default value of atol = 1e-07 and angles[1] with the default value of atol = 1e-06

Ideally for `angles[0], angles[2]`  and `angles[1]` we should separately search for tightest tolerance possible as described above.

For zero angles though we need to adjust `atol` instead of `rtol`. It might be a bit confusing, but I'm sure some better tolerance parameters in general can be found.

Reply:
My recipe:

- For `test_as_euler_compare_algorithms` search for smallest possible `rtol`, set `atol` to zero
- For `test_as_euler_degenerate_compare_algorithms`: do separate `assert_allclose` for each angle component. Without zero angles - search for `rtol` as in the previous item. With zero angles - first search for `rtol` as before, then search for `atol` (until the test pass)

Is it clear?

Reply:
Ok so I just did new changes.
Basically, you were right, only the zero angles for `angles[1]` are problematic.

For assymmetric sequences I compare `angles[0]` and `angles[2]` directly with `atol=0` and `atol=1e-12`, everything passes. For `angles[1]`, `atol=1e-8` gives an error, so I revert it back to `1e-7`.

For symmetric sequences, I calculate `angles[0]` and `angles[2]` directly with `atol=0` and `atol=1e-12`. Again, `angles[1]` (when different from 0), `atol=1e-8` gives an error, so I revert it back to `1e-7`. `angles[1]`, when zero, must have `atol=1e-6`

Reply:
For `test_as_euler_compare_algorithms`, using `n=10000`:

Everything passes with `atol=0`. 
For assymmetric axes, `rtol=1e-11` passes the test.
For symmetric axes, `rtol=1e-9` passes the test.